src/tarantool_schema.c

```diff
@@ -325,12 +349,75 @@ decode_index_parts_166(struct schema_field_value *parts, uint32_t part_count,
 		uint32_t len;
 		const char *str = mp_decode_str(data, &len);
 		part->field_type = field_type_by_name(str, len);
+		if (part->field_type == field_type_MAX)
+			return -1;
```

Comment:
When I looked into 'update processing the field type' commit, I think whether we should add such code here and in `parse_schema_space_value_value()`. My guess was that you intentionally don't do this in order to left the connector workable when tarantool has a space with, say, decimal or uuid index part.

Can you test it and clarify?

Reply:
Fixed

Reply:
We discussed it offline with Leonid and the fix is the following: all supported field types are added into the `enum field_type`.

It actually fix any possible 'unknown field type / index part type' problem during schema parsing with all tarantool versions that exist now. However I think it is not enough. We should also support future tarantool versions. I mean, when we'll introduce a new type in tarantool (say, `datetime`) and a user will use it for some space (as a field type in a format or as an index part type in an index), the connector should be able to make calls, evals and correctly work with other spaces.

I suggest to briefly look at https://github.com/tarantool/tarantool/issues/4632 — it is about quite similar problem, but with net.box client.

I can propose the following possible ways:

1. Don't store and don't validate data types (the idea appears when I [found](https://github.com/tarantool/tarantool-php/pull/164/files#r456669264) that our validation is not perfectly strict).
2. Store data types, but don't enforce them to be valid / known.
2.1. Exclude such space from a schema. Not good: misleading error at a request to the space.
2.2. Mark such space as unsupported in order to give a proper error.
2.3. Just save a type as `field_type_MAX` and consider a field type / index part type parsing as successful.

Both `1` and `2.3` are simple and powerful. `2.3` also adds a potential ability to use field types / index part types in a future to, say, expose a database schema to properly integrate with some data access framework / api. But `1` is simpler (and will give less memory footprint, but I think it is negligible).

There is point from @Leonid that strengthen `2.3` position: he wrote the schema parsing code quite similar to tarantool's one to easily update the connector when tarantool will be updated with some new schema parsing code.

Why I think that forward compatibility is important? Updating of a component in a project may require development resources and operation team resources (it very depends on details, but *some* resources are surely needed). Even just ensure that it is possible to downgrade a component back (if something going wrong after update), may require diving into the component details and changes details — it is just for example. The main idea is that an upgrade is not always cheap AFAIK and I don't want to make anybody's life harder. And, to be honest, I really like things, which just works without all those complex upgrade paths.

There is another point for the forward compatibility: it allows us to take a task of kind 'connector X does not support a data type Y' with less urgency if the connector still work in general.

I think that tarantool and its ecosystem is mature enough to take those problems seriously.

Reply:
Ok. I have implemented 2.3.